### ISSUE TITLE
Provide option to use alias method in multinomial sampling

### ISSUE THREAD

**Issue by neerajprad** [pytorch/pytorch#18906]
Ref to https://github.com/pytorch/pytorch/issues/4115. The alias method provides a very efficient way to sample from a categorical distribution in cases where the underlying distribution is static with many categories, and we want to sample from it many times. Once the alias table is built, sampling takes constant time w.r.t. the number of categories.

@vishwakftw has exposed the bindings to the sampler via `torch._multinomial_alias_setup` and `torch._multinomial_alias_draw` functions. We should make this available in `torch.distributions.categorical`. Since the `torch._multinomial_alias_setup` function is restricted to 1-D probability table, we cannot use this sampler with batched (and, distinct) `probs` argument.

My proposal would be to have the `Categorical.__init__` method take in an optional keyword argument `use_alias=False`. If the `probs/logits` have the correct shape, then we can construct and store the alias table. If `probs` is batched, we can raise a warning, set `use_alias=False` and use the default sampler. We will also need to make sure that `expand` works fine even if the `probs` tensor is subsequently expanded.

**Comment by rgemulla** [comment]
The following addition to pytorch may be enough for us:

https://github.com/pytorch/pytorch/issues/4115#issuecomment-471632622
